docs/architecture/psa-migration/strategy.md

```diff
@@ -256,6 +212,181 @@ Strategies currently (early 2022) used with each abstraction layer:
+`PSA_WANT_xxx` macros. For example, PSA-based code that want to use SHA-256
+will check for `PSA_WANT_ALG_SHA_256`, while legacy-based code that wants to
+use SHA-256 will check for `MBEDTLS_SHA256_C` if using the `mbedtls_sha256`
+API, or for `MBEDTLS_MD_C && MBEDTLS_SHA256_C` if using the `mbedtls_md` API.
+
+Code that obeys `MBEDTLS_USE_PSA_CRYPTO` will want to use one of the two
+dependencies above depending on whether `MBEDTLS_USE_PSA_CRYPTO` is defined:
+if it is, the code want the algorithm available in PSA, otherwise, it wants it
+available via the legacy API(s) is it using (MD and/or low-level).
+
+The strategy for steps 1 and 2 above will introduce new situations: code that
+currently compute hashes using MD (resp. a low-level hash module) will gain
+the ability to "fall back" to using PSA if the legacy dependency isn't
+available. Data related to a certain hash (OID, sizes, translations) should
+only be included in the build if it is possible to use that hash in some way.
+
+In order to cater to these new needs, new families of macros are introduced in
+`library/legacy_or_psa.h`, see its documentation for details.
+
+It should be noted that there are currently:
+- too many different ways of computing a hash (low-level, MD, PSA);
+- too many different ways to configure the library that influence which of
+  these ways is available and will be used (`MBEDTLS_USE_PSA_CRYPTO`,
+`MBEDTLS_PSA_CRYPTO_CONFIG`, `mbedtls_config.h` + `psa/crypto_config.h`).
+
+As a result, we need more families of dependency macros than we'd like to.
+This is a temporary situation until we move to a place where everything is
+based on PSA Crypto. In the meantime, long and explicit names where chosen for
+the new macros in the hope of avoiding confusion.
+
+Executing step 3 will mostly consist of using the right dependency macros in
+the right places (once the previous steps are done).
+
+**Note on testing**
+
+Since supporting driver-only builds is not about adding features, but about
+supporting existing features in new types of builds, testing will not involve
+adding cases to the test suites, but instead adding new components in `all.sh`
+that build and run tests in newly-supported configurations. For example, if
+we're making some part of the library work with hashes provided only by
+drivers when `MBEDTLS_USE_PSA_CRYPTO` is defined, there should be a place in
+`all.sh` that builds and run tests in such a configuration.
+
+There is however a risk, especially in step 3 where we change how dependencies
+are expressed (sometimes in bulk), to get things wrong in a way that would
+result in more tests being skipped, which is easy to miss. Care must be
+taken to ensure this does not happen. The following criteria can be used:
+
+- the sets of tests skipped in the default config and the full config must be
+  the same before and after the PR that implements step 3;
+- the set of tests skipped in the driver-only build is the same as in an
```

Comment:
I think that assuming the reviews are ideal, no such issue would happen. But sometimes, when renaming a lot of dependencies `en masse`, mistakes hard to spot might happen. Example:
A `.data` file test dependency on `MBEDTLS_SHA256_C` is bulk changed to `MBEDTLS_HAS_ALG_SHA_256_VIA_MD_OR_PSA`, since we aim to provide the same functionality via either of these configuration sets. Such rename is applied to 300 tests, and an initial check by the commiter doesn't show anything suspicious. He/she doesn't check the number of tests skipped.
As it turns out, one of the tests had a following dependency: `!MBEDTLS_SHA256_C` which was changed to `!MBEDTLS_HAS_ALG_SHA_256_VIA_MD_OR_PSA`. What if we didn't have the configuration combination with no MD and without PSA? Such test would be always skipped.
Maybe this exact example isn't accurate, since we have a build that would cover this case, but maybe there are some that we didn't check.
The idea behind this "skipped tests" script is to double check that there were no missed cases, regardless of the reason.

Reply:
Good point. Or: do a bulk renaming, then a merge/rebase brings in some new test cases for which the renaming wasn't applied.

Still, I don't understand why we'd want to do this comparison, as opposed to checking that there are no longer any old-style dependencies in places that should only have new-style dependencies.

Reply:
For example while working on `Driver-only hashes: X.509` I changed dependencies in `tests/suites/test_suite_x509write.data`. Replaced `MBEDTLS_MD5_C` and `MBEDTLS_SHAxxx_C` with the corresponding `MBEDTLS_HAS_ALG_xxx_VIA_MD_OR_PSA` macros.

Then when I run `component_test_psa_crypto_config_accel_hash_use_psa` from `all.sh` it turned out that set of `> Certificate write check Server1 SHA1` tests were skipped.
This was because there was also `MBEDTLS_SHA1_C` dependency in `.function` file.

It is easy to miss something. I found `outcome-analysis.sh` script very useful.